### PR TITLE
[Merged by Bors] - Respect alignment for zero-sized types stored in the world

### LINKED ISSUE #6615: `BlobVec` does not respect alignment for zero-sized types
## Bevy version

e48c05c73474a5f6a212257cecda7ceaf2465cf3

## Problem

Like `Vec`, `BlobVec` uses a dangling pointer as the backing storage for zero-sized types. However, `BlobVec` does not respect alignment, which results in instant UB when using a ZST with alignment other than 1.

```rust
#[test]
fn aligned_zst() {
    #[derive(Component)]
    #[repr(align(32))]
    struct Zst;

    let mut world = World::default();
    world.spawn(Zst);
    world.spawn(Zst);
    world.spawn(Zst);

    let mut q = world.query::<&Zst>();
    for &Zst in q.iter(&world) {}
}
```

[The offending line](https://github.com/bevyengine/bevy/blob/e48c05c73474a5f6a212257cecda7ceaf2465cf3/crates/bevy_ecs/src/storage/blob_vec.rs#L55): it creates a dangling pointer which is well-aligned for `u8`, not for the type being stored.

## Miri

```rust
error: Undefined Behavior: accessing memory with alignment 1, but alignment 32 is required
   --> C:\Users\joe10\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\slice\raw.rs:100:9
    |
100 |         &*ptr::slice_from_raw_parts(data, len)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ accessing memory with alignment 1, but alignment 32 is required
    |
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
    = note: BACKTRACE:
    = note: inside `std::slice::from_raw_parts::<'_, std::cell::UnsafeCell<tests::aligned_zst::Zst>>` at C:\Users\joe10\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\slice\raw.rs:100:9
note: inside `storage::blob_vec::BlobVec::get_slice::<tests::aligned_zst::Zst>` at crates\bevy_ecs\src\storage\blob_vec.rs:309:9
   --> crates\bevy_ecs\src\storage\blob_vec.rs:309:9
    |
309 |         std::slice::from_raw_parts(self.data.as_ptr() as *const UnsafeCell<T>, self.len)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `storage::table::Column::get_data_slice::<tests::aligned_zst::Zst>` at crates\bevy_ecs\src\storage\table.rs:191:9
   --> crates\bevy_ecs\src\storage\table.rs:191:9
    |
191 |         self.data.get_slice()
    |         ^^^^^^^^^^^^^^^^^^^^^
note: inside `<&tests::aligned_zst::Zst as query::fetch::WorldQuery>::set_table` at crates\bevy_ecs\src\query\fetch.rs:586:13
   --> crates\bevy_ecs\src\query\fetch.rs:586:13
    |
586 | /             table
587 | |                 .get_column(component_id)
588 | |                 .debug_checked_unwrap()
589 | |                 .get_data_slice()
    | |_________________________________^
note: inside `query::iter::QueryIterationCursor::<'_, '_, &tests::aligned_zst::Zst, ()>::next` at crates\bevy_ecs\src\query\iter.rs:595:21
   --> crates\bevy_ecs\src\query\iter.rs:595:21
    |
595 |                     Q::set_table(&mut self.fetch, &query_state.fetch_state, table);
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `<query::iter::QueryIter<'_, '_, &tests::aligned_zst::Zst, ()> as std::iter::Iterator>::next` at crates\bevy_ecs\src\query\iter.rs:53:13
   --> crates\bevy_ecs\src\query\iter.rs:53:13
    |
53  | /             self.cursor
54  | |                 .next(self.tables, self.archetypes, self.query_state)
    | |_____________________________________________________________________^
note: inside `tests::aligned_zst` at crates\bevy_ecs\src\lib.rs:87:21
   --> crates\bevy_ecs\src\lib.rs:87:21
    |
87  |         for &Zst in q.iter(&world) {}
    |                     ^^^^^^^^^^^^^^
note: inside closure at crates\bevy_ecs\src\lib.rs:76:5
   --> crates\bevy_ecs\src\lib.rs:76:5
    |
75  |       #[test]
    |       ------- in this procedural macro expansion
76  | /     fn aligned_zst() {
77  | |         #[derive(Component)]
78  | |         #[repr(align(32))]
79  | |         struct Zst;
...   |
87  | |         for &Zst in q.iter(&world) {}
88  | |     }
    | |_____^
    = note: this error originates in the attribute macro `test` (in Nightly builds, run with -Z macro-backtrace for more info)

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error

test tests::aligned_zst ... error: test failed, to rerun pass `-p bevy_ecs --lib`
```

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

Fixes #6615.

`BlobVec` does not respect alignment for zero-sized types, which results in UB whenever a ZST with alignment other than 1 is used in the world.

## Solution

Add the fn `bevy_ptr::dangling_with_align`.

---

## Changelog

+ Added the function `dangling_with_align` to `bevy_ptr`, which creates a well-aligned dangling pointer to a type whose alignment is not known at compile time.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Just so I understand the context more: why would you want to mess with the alignment of ZSTs? Definitely agree that this is needed, but I'm trying to piece this together better.

**Review by BoxyUwU — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

**Comment by JoJoJet** [comment]
> Just so I understand the context more: why would you want to mess with the alignment of ZSTs? Definitely agree that this is needed, but I'm trying to piece this together better.

I'm not aware of any actual use cases for this kind of alignment fiddling. This PR is about correctness more than anything, I would be surprised if this issue caused any problems in practice.

File: crates/bevy_ecs/src/storage/blob_vec.rs [hunk 1]
```diff
@@ -544,4 +546,26 @@ mod tests {
         // SAFETY: drop is able to drop a value of its `item_layout`
         let _ = unsafe { BlobVec::new(item_layout, Some(drop), 0) };
     }
+
+    #[test]
+    fn aligned_zst() {
```
Comment (james7132) [inline crates/bevy_ecs/src/storage/blob_vec.rs:551]: None of this actually tests the alignment of the ZST. Is that even possible?

A note should be made here that this is explicitly for UB checking with miri.

**Review by james7132 — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Merging: either we make a 0.9.1 release and it's in there, or we don't and there's no harm done by merging early.

bors r+

**Comment by JoJoJet** [comment]
So apparently this actually can matter in practice: https://doc.rust-lang.org/reference/type-layout.html

According to this, `[T; 0]` can have non-zero alignment on some platforms, so it's actually not impossible to cause this bug by accident. Granted, you'd need some *very* abstracted generic code to make that happen.

**Pull request closed without merging**